### PR TITLE
stability: deprecated API checks

### PR DESCRIPTION
**Pull request by deads2k**
Provides a sketch of how we could automatically create alerts and provide usage information about deprecated APIs.  They are starting to go away.

### CODE REVIEW

File: enhancements/kube-apiserver/stability-using-dead-api.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+1. Prevent deprecated APIs from being removed.
+2. This enhancement doesn't detail addition features we could add to upstream via a discovery doc and/or CRDs to make
+   deprecation more obvious and discoverable at runtime.
+   It simply works with what we have a available today.
+
+## Proposal
+
+1. Write a generator for alerts in kas-o (or openshift/kubernetes), that loads the current scheme (we functions for
+   this in openshift/api), navigates each type, determine which are deprecated and which are about to be removed.
+   This information can then be used to generate specific alerts and/or metrics.
```
Comment (mfojtik) [inline enhancements/kube-apiserver/stability-using-dead-api.md:68]: you still need to know if the deprecated types are in use, right? so you need to look into etcd?

File: enhancements/kube-apiserver/stability-using-dead-api.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+2. In the kas-o, if a "deprecated API used" alert has been fired, we have sufficient information to parse an
+   audit log and summarize the top X users of deprecated APIs.
```
Comment (mfojtik) [inline enhancements/kube-apiserver/stability-using-dead-api.md:70]: parsing audit log sounds pretty hacky to me, isn't there a better way? (admission/etc).

File: enhancements/kube-apiserver/stability-using-dead-api.md (hunk 1 above)
Comment (deads2k) [reply enhancements/kube-apiserver/stability-using-dead-api.md:68]: > you still need to know if the deprecated types are in use, right? so you need to look into etcd?

It's not about data in etcd, it's about how that data is accessed via the REST API.

File: enhancements/kube-apiserver/stability-using-dead-api.md (hunk 2 above)
Comment (deads2k) [reply enhancements/kube-apiserver/stability-using-dead-api.md:70]: > parsing audit log sounds pretty hacky to me, isn't there a better way? (admission/etc).

Admission doesn't cover get/list/watch.  I don't know of a better way to do it.  Perhaps a different audit sink, but that seems more fragile overall.

**Comment by mfojtik** [comment]
/lgtm

/hold
(for @openshift/openshift-team-openshift-architect approval)

**Comment by deads2k** [comment]
/lgtm cancel

I'd like to agree on who could do the work and have them fill in the rest.  Sort of like Luis did with the  p2p network detector

**Pull request closed without merging**